### PR TITLE
RAND: allow the primary, public and private of DRBGs to be overriden and replaced.

### PR DESCRIPTION
**Pull request by paulidale**
Allow the default DRBG set up to be modified in an application manner.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
@slontis, is this more like what you require?

File: doc/man3/RAND_get0_primary.pod [hunk 1]
```diff
@@ -28,6 +34,12 @@ directly, but is used internally to reseed the other two instances.
 
 These functions here provide access to the shared DRBG instances.
 
+The three functions all the default arrangement of DRBGs to be modified to
```
Comment (slontis) [inline doc/man3/RAND_get0_primary.pod:37]: You may want to reread this ..

**Review by slontis — COMMENTED** [review]
Please add a test, so we can see what this looks like.
A test that gets a known answer for RAND_priv_bytes_ex() in evp_libctx_test (this runs in both fips and non fips modes) should be sufficient..

File: doc/man3/RAND_get0_primary.pod (hunk 1 above)
Comment (paulidale) [reply doc/man3/RAND_get0_primary.pod]: What waffle I typed, and read and spell checked.
Rewritten now.

File: doc/man3/RAND_get0_primary.pod [hunk 2]
```diff
@@ -28,6 +34,13 @@ directly, but is used internally to reseed the other two instances.
 
 These functions here provide access to the shared DRBG instances.
 
+The three functions: RAND_set1_primary(),  RAND_set1_public() and
+RAND_set1_private() alter the default DRBGs to better suite an application's
+requirements.  In each case the specified EVP_RAND_CTX replaces any
+existing one and control is retained by the RAND subsystem.  Note that,
+the parentage of the replacement DRBG instances is arbitrary and need no be
```
Comment (slontis) [inline doc/man3/RAND_get0_primary.pod:41]: no=not

**Comment by paulidale** [comment]
Reference counting EVP_RAND_CTX's is something that will likely be required.

File: doc/man3/RAND_get0_primary.pod [hunk 3]
```diff
@@ -28,6 +34,13 @@ directly, but is used internally to reseed the other two instances.
 
 These functions here provide access to the shared DRBG instances.
 
+The three functions: RAND_set0_primary(),  RAND_set0_public() and
+RAND_set0_private() alter the default DRBGs to better suite an application's
```
Comment (mattcaswell) [inline doc/man3/RAND_get0_primary.pod:38]: Typo: s/suite/suit/

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by paulidale** [comment]
I'm unconvinced that this approach is a good one longer term.  It might be workable as an interim before 3.1 but that's about it.  There are several nasty gotchas.  Unfortunately, we've little time to figure the path out.

**Comment by mattcaswell** [comment]
> I'm unconvinced that this approach is a good one longer term. It might be workable as an interim before 3.1 but that's about it. There are several nasty gotchas. Unfortunately, we've little time to figure the path out.

So, are you withdrawing this PR? Or are we continuing with it? Why don't you think it is a good one?

**Comment by paulidale** [comment]
I'm not withdrawing it yet, I need to think it through further.

- Using the same EVP_RAND twice is a big problem but not disallowed.
- Using EVP_RANDs with a NULL parent will adversely impact system entropy.
- Switching EVP_RANDs could interfere with the reseeding.

We could write off such things as _user error_ but let's not lock ourselves into something that might go badly.

**Comment by t8m** [comment]
If in theory I have a FIPS validated RNG engine, can I still use it with OpenSSL (i.e. replace the default RAND method) and use the FIPS provider? If this is possible I would vote for postponing the replace-ability of primary/public/private DRBG to 3.1 at least.

**Comment by mattcaswell** [comment]
> If in theory I have a FIPS validated RNG engine, can I still use it with OpenSSL (i.e. replace the default RAND method) and use the FIPS provider?

IIUC, right now if you call RAND_bytes_ex() it will end up doing an `EVP_RAND_fetch` for an EVP_RAND called "CTR-DRBG" in the specified libctx. As long as you have a provider loaded that provides a "CTR-DRBG" (which both the default and fips providers do) then it will work. However you have no opportunity to replace it (using providers) with an alternative EVP_RAND algorithm. So for example, if you wanted to use "HASH-DRBG" or some custom DRBG then this can't be done.

You *can* still use the old RAND_METHOD way...but that's not a way we should be encouraging people to use.

**Comment by paulidale** [comment]
> If in theory I have a FIPS validated RNG engine, can I still use it with OpenSSL (i.e. replace the default RAND method) and use the FIPS provider? If this is possible I would vote for postponing the replace-ability of primary/public/private DRBG to 3.1 at least.

I believe that this is possible, when a EVP_RAND is created it is given a parent EVP_RAND which does not have to come from the same provider (or even library context) as the one being created.  FIPS either requires a special validation calling one a dependent module or having a secure channel (encrypted & authenticated) between the two.  The current reseeding setup (which is the only communication point) I believe meets the spirit of the second path.  No data leaves the FIPS boundaries, no code ever runs outside the FIPS boundaries so the channel could be considered secure (I'm not sure it actually meets the letter of the requirement, that's a lab qestion).

Specifically, get entropy calls to the parent do not involve the EVP layer or libcrypto.   I.e. a FIPS provider should be allowed to call a different FIPS provider for entropy and be okay.

**Comment by paulidale** [comment]
One complication with setting rand algorithm names early is due to their creation on demand.  They tend to be bought up very early on for quite a number of different reasons.  Being able to switch names early might not be enough -- 1.1.1 allows the DRBGs to be changed after being instantiated and it caused a nightmare.  This was _changing_ rather than _replacing_.

I'll keep thinking about this over the weekend and we can perhaps discuss options on Tuesday, if something reasonable appears possible in the time frame.  Otherwise, we can say for 3.1.

**Comment by paulidale** [comment]
This is going on a tangent we don't want to follow.  We want an ability to set a DRBG factory to create the public and private DRBGs and some way to create the primary (likely also a factory).  We'll also need to be able to set the type/falvour of each DRBG (or at least all of them together), possible the OSSL_PARAMs controlling them.  Locking/reference counting is also required (e.g. primary, public and private all pointing to one DRBG is legitimate).  Based on configuration files is also a positive.

Putting this back into draft mode while I mull the wine.

**Comment by mattcaswell** [comment]
> Putting this back into draft mode while I mull over the wine.

Have you finished drinking wine yet? What are your thoughts on this?

**Comment by paulidale** [comment]
I have a plan for what is required but implementation isn't remotely close yet.

- from this PR the set primary call
- different APIs for the public and private factories -- these need to either have enough information to create a new rand or take an (uninitialised) rand and ctx dup it.
- introduce EVP_RAND_CTX reference counting to avoid many badnesses

beyond this, these would be beneficial but possibly not required:
- configuration changes to allow the types of the the DRBGs to be specified via config rather than programmatically
- create a passthrough / null random
- separating out the CRNGT into its own rand (makes plumbing easier)
- separating out the entropy sources into their own rands (again plumbing, plus replacement)

The big confounding thing for beta is that I'm not sure any of this is strictly necessary -- the lab can do their required testing without any of it (albeit it will be more difficult and uglier) and users' haven't lost anything, so this is more a feature.

**Comment by slontis** [comment]
Pauli - I think we should address the property query initialization problem and leave the rest alone for 3.0.

**Comment by richsalz** [comment]
Have you confirmed that the lab doesn't need this? If that is the case, then I would really rather see nothing done here, for a couple of reasons:
- There are many "outside" PR's that still need review and perhaps merging.
- The RAND DRBG etc API's are not good, and adding more things to that base without a complete rethink will not help.

**Comment by slontis** [comment]
> Have you confirmed that the lab doesn't need this? If that is the case, then I would really rather see nothing done here, for a couple of reasons:

In cases where the lab has needed this in the past, we should be able to inject the random in via parameters (i.e set_params). I have already done this in some places where I know it is needed  - but it is more than possible that I have missed something (We will have to deal with that if it happens - it is not hard to add).
I would prefer to do that than add a RAND function that we will be stuck with..

**Comment by mattcaswell** [comment]
Ping! What is happening with this? Are we pursuing it or closing it? We need to make a decision asap.

**Comment by paulidale** [comment]
If #12931 is accepted, I don't think this is required.

**Comment by mattcaswell** [comment]
In that case should we put that one in the 3.0 beta 1 milestone, and this PR out of it?

**Comment by mattcaswell** [comment]
> In that case should we put that one in the 3.0 beta 1 milestone, and this PR out of it?

I've made that milestone change.

**Comment by paulidale** [comment]
Yes, this isn't for beta.  The other should be.

**Pull request closed without merging**